pkg/avx/collector.go

```diff
@@ -77,111 +102,126 @@ var (
+	if err != nil {
+		return -1, errors.Wrap(err, "unable to read tracepoint ID")
+	}
+
+	tid, err := strconv.Atoi(strings.TrimSpace(string(id)))
 	if err != nil {
-		return 0, 0, 0, errors.Wrapf(err, "unable to open ELF file %s", path)
+		return -1, errors.New("unable to convert tracepoint ID")
 	}
-	defer elfFile.Close()
 
-	sec := elfFile.Section("version")
-	if sec == nil {
-		return 0, 0, 0, errors.New("unable to find 'version' section")
+	attr := unix.PerfEventAttr{
+		Type:        unix.PERF_TYPE_TRACEPOINT,
+		Config:      uint64(tid), // tracepoint id
+		Sample_type: unix.PERF_SAMPLE_RAW,
+		Sample:      1,
+		Wakeup:      1,
 	}
-	data, err := sec.Data()
+
+	pfd, err := unix.PerfEventOpen(&attr, -1, 0, -1, unix.PERF_FLAG_FD_CLOEXEC)
 	if err != nil {
-		return 0, 0, 0, errors.Wrap(err, "unable to get version data")
+		return -1, errors.Wrap(err, "unable to open perf events")
 	}
 
-	// Least Significant Byte first
-	return data[2], data[1], data[0], nil
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_ENABLE, 0); errno != 0 {
+		return -1, errors.Errorf("unable to set up perf events: %s", errno)
+	}
+
+	if _, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(pfd), unix.PERF_EVENT_IOC_SET_BPF, uintptr(prog.FD())); errno != 0 {
+		return -1, errors.Errorf("unable to attach bpf program to perf events: %s", errno)
+	}
+
+	return pfd, nil
 }
 
-func checkElfKernelVersion(path string) error {
-	elfMajor, elfMinor, _, err := getElfKernelVersion(path)
+func getKernelVersion() uint32 {
+
+	var uts unix.Utsname
+
+	err := unix.Uname(&uts)
 	if err != nil {
-		return err
+		return 0
 	}
 
-	currentCode, err := bpf.CurrentKernelVersion()
+	str := string(bytes.SplitN(uts.Release[:], []byte{0}, 2)[0])
+
+	ver := strings.SplitN(str, ".", 3)
+
+	major, err := strconv.ParseUint(ver[0], 10, 8)
 	if err != nil {
-		return errors.Wrap(err, "unable to get current kernel version")
+		return 0
 	}
-
-	if currentCode < kernelVersionCode(elfMajor, elfMinor, 0) {
-		return errors.New("host kernel is too old, consider rebuilding eBPF")
+	minor, err := strconv.ParseUint(ver[1], 10, 8)
+	if err != nil {
+		return uint32(major << 16)
 	}
 
-	return nil
+	// ignore patch version
+	return uint32(major<<16 + minor<<8)
 }
 
-type collector struct {
-	root                     string
-	elfFilepath              string
-	bpfModule                *bpf.Module
-	avxContextSwitchCounters *bpf.Map
-	allContextSwitchCounters *bpf.Map
-	lastUpdateNs             *bpf.Map
-	lastCPUCounters          *bpf.Map
+func kernelVersionStr(v uint32) string {
+	return fmt.Sprintf("%d.%d.0", v>>16, (v>>8)&0xff)
 }
 
 // NewCollector creates new Prometheus collector for AVX metrics
 func NewCollector() (prometheus.Collector, error) {
 
-	elfFilepath := path.Join(bpfInstallpath, bpfBinaryName)
-
-	if err := checkElfKernelVersion(elfFilepath); err != nil {
-		return nil, err
+	// Set rlimit to be able to lock map values in memory
+	memlockLimit := &unix.Rlimit{
```

Comment:
I think we can do this here for now, but in the long run I think the process limits, caps, etc. should come from systemd service files or similar. This might be easier said than done though. :-)

Reply:
Right. This used to be "unlimited" but I changed it to minimum needed.